libpod/runtime_pod_namespaces_linux.go

```diff
@@ -0,0 +1,88 @@
+// +build linux
+
+package libpod
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/docker/docker/pkg/mount"
+	"github.com/docker/docker/pkg/symlink"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+	"k8s.io/release/pkg/command"
+)
+
+func (r *Runtime) pinNamespaces(p *Pod) error {
+	logrus.Debugf("pinning namespaces")
+
+	// TODO: make those configurable
+	namespacesDir := "/var/run/libpod/ns"
+	pinns := "pinns"
+
+	pinDir := filepath.Join(namespacesDir, p.ID())
+	if err := os.MkdirAll(pinDir, 0755); err != nil {
+		return errors.Wrap(err, "unable to create namespaces dir")
+	}
+
+	args := []string{"--dir", pinDir}
+	// Not supported by cri-o/pinns yet, but by pinns.rs
+	/* if p.config.UsePodCgroup {
+		args = append(args, "--cgroup")
+	} */
+	if p.config.UsePodIPC {
+		args = append(args, "--ipc")
+	}
+	if p.config.UsePodNet {
+		args = append(args, "--net")
+	}
+	// Not supported by cri-o/pinns yet, but by pinns.rs
+	/* if p.config.UsePodPID {
+		args = append(args, "--pid")
+	} */
+	if p.config.UsePodUTS {
+		args = append(args, "--uts")
+	}
+
+	if err := command.New(pinns, args...).RunSilentSuccess(); err != nil {
```

Comment:
Can we use the stdlib instead?

Reply:
Thinking about it: couldn't we do everything in a go package with C bindings? `github.com/containers/libpod/pkg/pinns`? That would be faster, easier to maintain (as most folks don't speak rust unfortunately), and there would be no need to package a new tool for all distributions.

Reply:
cgo isn't a great option, because the go runtime and locking threads is funky. We thought about this approach, but ultimately decided exec'ing the binary was a more guarenteed way of preventing an interrupt from changing the context too much. here's more info https://github.com/containernetworking/plugins/tree/master/pkg/ns

Reply:
That's an omnipresent problem in go but locking the thread has proven to work. I didn't yet buy the idea of adding another dependency on top.

Reply:
Just saw https://github.com/cri-o/cri-o/tree/master/pinns now. Easier than Rust. I try to catch up with the discussion in the CRI-O PR.

Reply:
Don’t take my Rust stuff too serious, we had a hackweek and I’m totally open to stick to the C version if that’s the preferred one. 🙂 

I was at the same point in thinking to unshare+pin as subcommand rather than having another binary, but now I see totally the arguments against something like that. 🤔

Which kinds of namespaces do we want to unshare beside NET, IPC and UTS? What about PID (dedicated feature called “pns“ in k8s), CGROUP and the USER (permissions) ns?